sdk/testing/stepwise/stepwise.go

```diff
@@ -0,0 +1,345 @@
+// Package stepwise offers types and functions to enable black-box style tests
+// that are executed in defined set of steps. Stepwise utilizes "Drivers" which
+// setup a running instance of Vault and provide a valid API client to execute
+// user defined steps against.
+package stepwise
+
+import (
+	"fmt"
+	"os"
+	"testing"
+
+	log "github.com/hashicorp/go-hclog"
+
+	"github.com/hashicorp/vault/api"
+	"github.com/hashicorp/vault/sdk/helper/consts"
+	"github.com/hashicorp/vault/sdk/helper/logging"
+)
+
+// StepOperation defines operations each step could preform
+type StepOperation string
+
+const (
+	WriteOperation  StepOperation = "create"
+	UpdateOperation               = "update"
+	ReadOperation                 = "read"
+	DeleteOperation               = "delete"
+	ListOperation                 = "list"
+	HelpOperation                 = "help"
+)
+
+// StepDriver is the interface Drivers need to implement to be used in
+// Case to execute each Step
+type StepDriver interface {
+	Setup() error
+	Client() (*api.Client, error)
+	Teardown() error
+	Name() string // maybe?
+
+	// ExpandPath adds any Namespace or mount path to the user defined path
+	ExpandPath(string) string
+
+	// MountPath returns the path the plugin is mounted at
+	MountPath() string
+
+	// BarrierKeys returns the keys used to seal/unseal the cluster. Used for
+	// debugging. TODO verify we should provide this
+	//BarrierKeys()        [][]byte
+
+	// RootToken returns the root token of the cluster, used for making requests
+	// as well as administrative tasks
+	RootToken() string
+}
+
+// PluginType defines the types of plugins supported
+// This type re-create constants as a convienence so users don't need to import/use
+// the consts package.
+type PluginType consts.PluginType
+
+// These are originally defined in sdk/helper/consts/plugin_types.go
+const (
+	PluginTypeUnknown PluginType = iota
+	PluginTypeCredential
+	PluginTypeDatabase
+	PluginTypeSecrets
+)
+
+func (p PluginType) String() string {
+	switch p {
+	case PluginTypeUnknown:
+		return "unknown"
+	case PluginTypeCredential:
+		return "auth"
+	case PluginTypeDatabase:
+		return "database"
+	case PluginTypeSecrets:
+		return "secret"
+	default:
+		return "unsupported"
+	}
+}
+
+// DriverOptions are a collection of options each step driver should
+// support
+type DriverOptions struct {
```

Comment:
This is good feedback, and in my mind I've imagined that we'd likely need to eventually support passing in some kind of custom configuration that's specific to the `driver` (now called `environment`). I suspect it would be some kind of `map[string]interface{}` that the environment will need to know how to parse. 

For now I'm going to leave this as is and think more on it. The use of the options is more apparent in the example tests which will be shared in a future PR when the first environment is added.

Reply:
> in my mind I've imagined that we'd likely need to eventually support passing in some kind of custom configuration that's specific to the `driver` (now called `environment`).

Yep, I see the following cases:

* Environment supports exactly what specified in `EnvironmentOptions`
* Environment supports what in `EnvironmentOptions` + some extra environment-specific options
* Environment supports less than what specified in `EnvironmentOptions`.

Given the above and the fact that environment is going to be instantiated in plugin's test itself (like in [`TestBackend_basic_docker`, for example](https://github.com/hashicorp/vault/blob/5b8f2f71dbd1fbb4291cb90ba4048b3f817a4f0b/builtin/logical/transit/stepwise_test.go#L18-L40)), I think it makes sense to let environments to define their own options and the way how they accept options (constructor arguments for each option, a struct for everything or functional options as @pcman312 suggested or whathever). 

This way we will keep the framework a bit simpler: it will rely on the interface only and will not care about how an envirionment was instantiated.

Reply:
I renamed `DriverOptions` to `EnvironmentOptions`, and then later simply `MountOptions`. The intent for this type is to configure how the plugin is mounted, specifically the path. `Driver` was outdated and `Environment` was a misnomer, so I changed to `MountOptions`. 

Going forward, if it's appropriate to enable developers to specify environment specific options, I think the functional options method suggested makes the most sense. For now though, that feature is deferred until we know more about how it would be used.